packages/mephisto-remote-query/src/index.js

```diff
@@ -0,0 +1,154 @@
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React from "react";
+
+import ConnectionIndicator from "./ConnectionIndicator.jsx";
+import {
+  useMephistoLiveTask,
+  AGENT_STATUS,
+  STATUS_TO_TEXT_MAP,
+  MephistoContext,
+  ErrorBoundary,
+} from "mephisto-task";
+
+// Generate a random id
+function uuidv4() {
+  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function (c) {
+    var r = (Math.random() * 16) | 0,
+      v = c == "x" ? r : (r & 0x3) | 0x8;
+    return v.toString(16);
+  });
+}
+
+const useMephistoRemoteQueryTask = function (props) {
+  // mapping of request ID to the callback object
+  const [requestCallbacks, addRequestCallback] = React.useReducer(
+    (previousRequestCallbacks, newRequestCallback) => {
+      let requestId = newRequestCallback.request_id;
+      let newRequestCallbacks = {
+        ...previousRequestCallbacks,
+        [requestId]: newRequestCallback,
+      };
+      return newRequestCallbacks;
+    },
+    {}
+  );
+
+  const [disconnectIssueText, setDisconnectIssueText] = React.useState();
+
+  // We register a ref so that the liveUpdate handler has access
+  // to the current set of callbacks
+  const requestCallbacksRef = React.useRef();
+  requestCallbacksRef.current = requestCallbacks;
+
+  function handleRemoteResponse(liveUpdate) {
+    let targetRequest = requestCallbacksRef.current[liveUpdate.handles];
+    if (targetRequest === undefined) {
+      console.log("No request found to handle this live update", liveUpdate);
+      return;
+    }
+
+    let parsedResponse = JSON.parse(liveUpdate.response);
+    targetRequest.callback({
+      requestId: targetRequest.request_id,
+      response: parsedResponse,
+      args: targetRequest.args,
+    });
+  }
+
+  let mephistoProps = useMephistoLiveTask({
+    onStatusUpdate: ({ status }) => {
+      if (
+        [
+          AGENT_STATUS.DISCONNECT,
+          AGENT_STATUS.RETURNED,
+          AGENT_STATUS.EXPIRED,
+          AGENT_STATUS.TIMEOUT,
+          AGENT_STATUS.MEPHISTO_DISCONNECT,
+        ].includes(status)
+      ) {
+        setDisconnectIssueText(STATUS_TO_TEXT_MAP[status]);
+      }
+    },
+    onLiveUpdate: (liveUpdate) => {
+      handleRemoteResponse(liveUpdate);
+    },
+  });
+
+  // We remove the live socket props, as we intend to provide a different interface
+  let {
+    connect,
+    destroy,
+    sendLiveUpdate,
+    ...otherMephistoProps
+  } = mephistoProps;
+  let { agentId } = mephistoProps;
+
+  // They're still exposed in a separate manner though
+  const _fullSocketProps = {
+    connect,
+    destroy,
+    sendLiveUpdate,
+  };
```

Comment:
what if someone wants to use remote procedure calls, while also having a live task. any reason we shouldn't just support that? this means that presenting remote calls as another tool in the toolbelt, instead of a separate modality (different toolbelt altogether).

what this means in practice is something like `useMephistoLiveTask` just exposing a `makeRemoteCall` method, that exists in case people want to use it. is there something i'm missing there, maybe around packets not being differentiated correctly if they're traveling over the same channel?

Reply:
> maybe around packets not being differentiated correctly if they're traveling over the same channel?

Yeah this is the exact issue, the current setup requires that people use an extension of the `RemoteQueryBlueprint` (or something that implements its API) in order to properly use this package, as that blueprint converts the `liveUpdate` format to be for RPCs. `useMephistoLiveTask` is a strict superset of this functionality, so I'm not sure what the use case would be for allowing both at the same time.